internal/source/kubernetes/source.go

```diff
@@ -184,78 +241,119 @@ func consumeEvents(ctx context.Context, s Source) {
 	for _, eventType := range eventTypes {
 		router.RegisterEventHandler(
 			ctx,
-			s,
 			eventType,
-			handleEvent,
+			s.handleEventFn(globalLogger),
 		)
 	}
 
 	router.HandleMappedEvent(
 		ctx,
-		s,
 		config.ErrorEvent,
-		handleEvent,
+		s.handleEventFn(globalLogger),
 	)
 
+	globalLogger.Info("Starting background process...")
 	stopCh := ctx.Done()
 	dynamicKubeInformerFactory.Start(stopCh)
+	<-stopCh
+	dynamicKubeInformerFactory.Shutdown()
+	globalLogger.Info("Stopped background process...")
+	return nil
 }
 
-func handleEvent(ctx context.Context, s Source, e event.Event, updateDiffs []string) {
-	s.logger.Debugf("Processing %s to %s/%v in %s namespace", e.Type, e.Resource, e.Name, e.Namespace)
-	enrichEventWithAdditionalMetadata(s, &e)
+func (s *Source) handleEventFn(log logrus.FieldLogger) func(ctx context.Context, e event.Event, sources, updateDiffs []string) {
+	globalLogger := log
 
-	// Skip older events
-	if !e.TimeStamp.IsZero() && e.TimeStamp.Before(s.startTime) {
-		s.logger.Debug("Skipping older events")
-		return
-	}
+	return func(ctx context.Context, e event.Event, sources, updateDiffs []string) {
+		globalLogger.Debugf("Processing %s to %s/%v in %s namespace", e.Type, e.Resource, e.Name, e.Namespace)
 
-	// Check for significant Update Events in objects
-	if e.Type == config.UpdateEvent && len(updateDiffs) > 0 {
-		e.Messages = append(e.Messages, updateDiffs...)
-	}
+		// Skip older events
+		if !e.TimeStamp.IsZero() && e.TimeStamp.Before(s.bgProcessor.StartTime()) {
+			globalLogger.Debug("Skipping older event...")
+			return
+		}
 
-	// Filter events
-	e = s.filterEngine.Run(ctx, e)
-	if e.Skip {
-		s.logger.Debugf("Skipping event: %#v", e)
-		return
-	}
+		// Check for significant Update Events in objects
+		if e.Type == config.UpdateEvent && len(updateDiffs) > 0 {
+			e.Messages = append(e.Messages, updateDiffs...)
+		}
 
-	if len(e.Kind) <= 0 {
-		s.logger.Warn("sendEvent received e with Kind nil. Hence skipping.")
-		return
-	}
+		errs := multierror.New()
+		for _, sourceKey := range sources {
+			eventCopy := e
 
-	recRunner, recCfg := s.recommFactory.New(s.config)
-	err := recRunner.Do(ctx, &e)
-	if err != nil {
-		s.logger.Errorf("while running recommendations: %w", err)
-		return
-	}
+			srcCfg, ok := s.configStore.Get(sourceKey)
+			if !ok {
+				errs = multierror.Append(errs, fmt.Errorf("source with key %q not found", sourceKey))
+			}
 
-	if recommendation.ShouldIgnoreEvent(&recCfg, e) {
-		s.logger.Debugf("Skipping event as it is related to recommendation informers and doesn't have any recommendations: %#v", e)
-		return
-	}
+			if srcCfg.ActiveSourceConfig == nil {
+				globalLogger.Errorf("ActiveSourceConfig not found for source %s", srcCfg.name)
```

Comment:
if error should we then add it to `errs`? from what I see, it should never happen, right? so if it happens, then there is a bug?

Reply:
yes, it shouldn't happen. Agree, should be added to errs.
I've replaced the line with

```go
errs = multierror.Append(errs, fmt.Errorf("ActiveSourceConfig not found for source %s. This seems to be a bug", srcCfg.name))
```